src/encodings/levels.rs

```diff
@@ -25,52 +25,70 @@ use errors::{ParquetError, Result};
 use util::bit_util::{ceil, log2, BitReader, BitWriter};
 use util::memory::ByteBufferPtr;
 
-enum InternalEncoder {
-  RLE(RleEncoder),
-  BIT_PACKED(BitWriter)
-}
-
-enum InternalDecoder {
-  RLE(RleDecoder),
-  BIT_PACKED(BitReader)
+/// Computes max buffer size for level encoder/decoder based on encoding, max
+/// repetition/definition level and number of total buffered values (includes null
+/// values).
+#[inline]
+pub fn max_buffer_size(
+  encoding: Encoding,
+  max_level: i16,
+  num_buffered_values: usize
+) -> usize {
+  let bit_width = log2(max_level as u64 + 1) as u8;
+  match encoding {
+    Encoding::RLE => {
+      RleEncoder::max_buffer_size(bit_width, num_buffered_values) +
+        RleEncoder::min_buffer_size(bit_width)
+    },
+    Encoding::BIT_PACKED => {
+      ceil((num_buffered_values * bit_width as usize) as i64, 8) as usize
+    },
+    _ => panic!("Unsupported encoding type {}", encoding)
+  }
 }
 
-/// A encoder for definition/repetition levels.
-/// Currently only supports RLE and BIT_PACKED (dev/null) encoding.
-pub struct LevelEncoder {
-  bit_width: u8,
-  encoder: InternalEncoder
+/// Encoder for definition/repetition levels.
+/// Currently only supports RLE and BIT_PACKED (dev/null) encoding, including v2.
```

Comment:
what does `dev/null` here mean?

Reply:
I do not quite remember, it was something similar to no-op encoding. I will remove it.